doc/design/zarr-support.md

```diff
@@ -0,0 +1,100 @@
+3. Blob url should continue to point to a single location (i.e. the prefix ending with `/` common to all keys for zarr files) in an s3 bucket instead of a collection of files.
+4. The `etag` connected with the zarr file should be a tree-hash of dandi-etags of the folder.
+
+Storing each sub-file of a zarr "dataset" as an asset seems like an overkill without any specific gain at this point.
+
+Given these considerations here are questions for implementation
+1. Is there a way to upload a folder to a given prefix using an API key without having to create 100k signed URLs?
+1. Should the tree structure be stored somewhere so that diffs can be ascertained?
+   * Assuming 64 bytes of checksum, 64 bytes for path (optimistic), and 1million files (pessimistic), that's 128MB of checksum info in the DB. Plausible for upload, but probably not long term.
+1. Given that each zarr file may contain 100k+ files, how will dandi-cli handle alterations?
+
+# Requirements
+
+1. zarr files are stored in a "directory" in S3.
+1. Each zarr file corresponds to a single Asset.
+1. The CLI uses some kind of tree hashing scheme to compute a checksum for the entire zarr file. The API verifies this checksum _immediately_ after upload; it's not good enough to download the entire zarr file to calculate it after upload.
+1. The system can theoretically handle zarr files with ~1 million subfiles, each of size 64 * 64 * 64 bytes ~= 262 kilobytes.
+
+# Implementation
+
+## Hash function for checksums
+The hash function used to generate the checksum for the zarr file must have these characteristics:
+
+1. It must be based on the dandi-etag function of the subfile contents, and the path of the subfile in the zarr file.
+1. A well-defined ordering of the subfiles in the zarr file must exist.
+The checksum must be computed on the subfiles in this order.
+1. It must be applicable on one subfile at a time.
+1. It must be able save its state between subfiles.
+
+**TODO** what is this function?
+
+A simple scheme that I think would work:
+
+* initial value is `sha256("{etag}:{path}")` for the first subfile
+* the next value is `sha256("{prev_sha256}:{etag}:{path}")`, ad infinitum
+
+We could also do something more complicated with merkle trees or something, but it would be much more obnoxious to compute.
+
+## Upload flow
+
+### Before upload (technically optional)
+1. CLI calculates the checksum of the zarr file.
+1. CLI queries the API (**TODO** URL?) if the checksum has already been uploaded.
+1. If so, it proceeds with the already uploaded zarr file and skips upload entirely.
+
+### Upload
+1. CLI queries the API (**TODO** URL?) with the checksum of the zarr file to initiate an upload. The API creates an upload UUID, records the checksum, and initializes a "running checksum" to `null`.
+1. CLI requests a batch of presigned URLs (**TODO** URL?).
+The files must be ordered in the same way used to calculate the checksum.
```

Comment:
I thought that idea was then to operate via providing IAM credentials with write access to target upload prefix.
If that is the case, no need for presigned URLs, but rather instead of `parts` in response provide IAM credentials (access key, secret key, access token, expiration).  What we might need though is a dedicated API endpoint to provide renewed credentials if prior ones are (about to) expire(d).

Reply:
If we were to operate on per-file upload basis (thus ridiculous amount of presigning) and provide this list -- I would not rely on "ordered in the same way". It should be explicitly ordered by checksumming  (we need to assume that CLI and dandi-api both use the same algorithm) which would sort (or not) internally to that algorithm consistently. We should not rely on external to algorithm sorting to provide that sorted order to both checksum compute and this upload end point.

Reply:
This design doesn't involve IAM credentials. If it doesn't solve the requirements, we can throw out this plan and come up with something else, possibly involving IAM.

If possible, I would rather avoid having to do all that IAM management. It's a technically viable option, but I don't thing we _have_ to do it that way.

Reply:
I will stop further analysis below since it roots at the idea of "presigning" and I probably incorrectly thought that we would like to avoid that for 100000s of files zarr will be.  Please clarify @dchiquito and @satra

Reply:
After consideration my opinion is that presigning is the least bad thing. It involves `number of files / files per batch` requests, but ultimately that's much less data than fully proxying the entire upload. Any direct upload scheme like the IAM idea has to figure out a way to calculate the checksum.

I think this scheme I laid out meets the requirements. For now it would be helpful to identify any shortcomings it has (I listed a few in the Pros/Cons section) so we can modify it or discard it in favor of something else.